File: 4 Algorithms on Strings/Week 4 Constructing Suffix Arrays and Suffix Trees/Programming-Assignment-3/suffix_array_matching/suffix_array_matching.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <cmath>

using namespace std;
const int maxn = 2 * 1e5 + 5;
const int alphabet = 256;
vector<int> order(maxn), classes(maxn);
set<int> occurrences;
string s, pat;
int n, m;

vector<int> sortCharacters() {
  vector<int> order(maxn), cnt(alphabet);
  for (int i = 0; i < n; ++i)
    cnt[s[i]]++;
  for (int i = 1; i < alphabet; ++i)
    cnt[i] += cnt[i - 1];
  for (int i = n - 1; i >= 0; --i) {
    int c = s[i];
    cnt[c]--;
    order[cnt[c]]= i;
  }
  return order;
}

vector<int> computeCharClasses(const vector<int>& order) {
  vector<int> classes(maxn);
  classes[order[0]] = 0;
  for (int i = 1; i < n; ++i) {
    if (s[order[i]] != s[order[i - 1]])
      classes[order[i]] = classes[order[i - 1]] + 1;
    else
      classes[order[i]] = classes[order[i - 1]];
  }
  return classes;
}

vector<int> sortDoubled(int l, const vector<int>& order, const vector<int>& classes) {
  vector<int> cnt(n), newOrder(maxn);
  for (int i = 0; i < n; ++i)
    cnt[classes[i]]++;
  for (int i = 1; i < n; ++i)
    cnt[i] += cnt[i - 1];
  for (int i = n - 1; i >= 0; --i) {
    int start = (order[i] - l + n) % n;
    int c = classes[start];
    cnt[c]--;
    newOrder[cnt[c]] = start;
  }
  return newOrder;
}

vector<int> updateClasses(int l, const vector<int>& newOrder, const vector<int>& classes) {
  vector<int> newClasses(maxn);
  newClasses[newOrder[0]] = 0;
  for (int i = 1; i < n; ++i) {
    int cur = newOrder[i];
    int prev = newOrder[i - 1];
    int mid = (cur + l) % n;
    int midPrev = (prev + l) % n;
    if (classes[cur] != classes[prev] || classes[mid] != classes[midPrev])
      newClasses[cur] = newClasses[prev] + 1;
    else
      newClasses[cur] = newClasses[prev];
  }
  return newClasses;
}

vector<int> buildSuffixArray() {
  order = sortCharacters();
  classes = computeCharClasses(order);
  int l = 1;
  while (l < n) {
    order = sortDoubled(l, order, classes);
    classes = updateClasses(l, order, classes);
    l *= 2;
  }
  return order;
}

void patternMatching(char* s, char* pat, const vector<int>& suffix_array, int m) {
  int l = 0, r = n, start, end;
  while (l < r) {
    int mid = (l + r) / 2;
    int res = strncmp(pat, s + suffix_array[mid], m);
    if (res > 0) l = mid + 1;
    else r = mid;
  }
  start = l;
  r = n;
  while (l < r) {
    int mid = (l + r) / 2;
    int res = strncmp(pat, s + suffix_array[mid], m);
    if (res < 0) r = mid;
    else l = mid + 1;
  }
  end = r;
  if (start <= end)
    for (int i = start; i < end; ++i)
      occurrences.insert(suffix_array[i]);
}

int main() {
  cin >> s >> m;
  s += '$';
  n = s.size();

  vector<int> suffix_array = buildSuffixArray();

  for (int i = 0; i < m; ++i) {
    cin >> pat;
    patternMatching(&s[0], &pat[0], suffix_array, pat.size());
  }

  for (auto x: occurrences)
    cout << x << ' ';
  cout << endl;

  return 0;
}

File: 4 Algorithms on Strings/Week 4 Constructing Suffix Arrays and Suffix Trees/Programming-Assignment-3/suffix_array_matching/suffix_array_matching.py
# python3
import sys

def SortCharacters(S):
    order = [0] * len(S)
    Sigma = sorted(set(S))
    count = [0] * len(Sigma)
    
    for i in range(len(S)):
        idx = Sigma.index(S[i])
        count[idx] = count[idx] + 1
    for j in range(1, len(Sigma)):
        count[j] = count[j] + count[j - 1]
    for i in range(len(S) - 1, 0, -1):
        c = S[i]
        idx = Sigma.index(c)
        count[idx] = count[idx] - 1
        order[count[idx]] = i
    return order
        
def ComputeCharClasses(S, order):
    clss = [0] * len(S)
    for i in range(1, len(S)):
        if S[order[i]] != S[order[i - 1]]:
            clss[order[i]] = clss[order[i - 1]] + 1
        else:
            clss[order[i]] = clss[order[i - 1]]
    return clss
    
def SortDoubled(S, L, order, clss):
    count = [0] * len(S)
    newOrder = [0] * len(S)
    for i in range(len(S)):
        count[clss[i]] = count[clss[i]] + 1
    for j in range(1, len(S)):
        count[j] = count[j] + count[j - 1]
    for i in range(len(S) - 1, -1, -1):
        start = (order[i] - L + len(S)) % len(S)
        cl = clss[start]
        count[cl] = count[cl] - 1
        newOrder[count[cl]] = start
    return newOrder

def UpdateClasses(newOrder, clss, L):
    n = len(newOrder)
    newClass = [0] * n
    
    for i in range(1, n):
        cur = newOrder[i]
        prev = newOrder[i - 1]
        mid = (cur + L)
        midPrev = (prev + L) % n
        if clss[cur] != clss[prev] or clss[mid] != clss[midPrev]:
            newClass[cur] = newClass[prev] + 1
        else:
            newClass[cur] = newClass[prev]
    return newClass
            
def build_suffix_array(text):
  """
  Build suffix array of the string text and
  return a list result of the same length as the text
  such that the value result[i] is the index (0-based)
  in text where the i-th lexicographically smallest
  suffix of text starts.
  """
  order = SortCharacters(text)
  clss = ComputeCharClasses(text, order)
  L = 1
  while L < len(text):
      order = SortDoubled(text, L, order, clss)
      clss = UpdateClasses(order, clss, L)
      L = 2 * L
  return order

"""
Use PatternMatchingWithSuffixArray Algorithm in Week 2
"""
def find_occurrences(text, patterns):
    text += '$'
    occs = []
    SuffixArray = build_suffix_array(text)
    
    for pattern in patterns:
        minIndex = 0
        maxIndex = len(text)
        while minIndex < maxIndex:
            midIndex = (minIndex + maxIndex) / 2
            if pattern > text[SuffixArray(midIndex):]:
                minIndex = midIndex + 1
            else:
                maxIndex = midIndex
        start = minIndex
        maxIndex = len(text)
        while minIndex < maxIndex:
            midIndex = (minIndex + maxIndex) / 2
            if pattern < text[SuffixArray(midIndex):]:
                maxIndex = midIndex
            else:
                minIndex = midIndex + 1
        end = maxIndex
        if start <= end:
            for i in range(start, end):
                occs.append(SuffixArray[i])
            
    return occs

if __name__ == '__main__':
    text = sys.stdin.readline().strip()
    pattern_count = int(sys.stdin.readline().strip())
    patterns = sys.stdin.readline().strip().split()
    occs = find_occurrences(text, patterns)
    print(" ".join(map(str, occs)))